indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/Asn1Encoding.kt

```diff
@@ -518,3 +510,48 @@ fun ByteArray.ensureSize(size: Int): ByteArray = (this.size - size).let { toDrop
 
 @Suppress("NOTHING_TO_INLINE")
 inline fun ByteArray.ensureSize(size: UInt) = ensureSize(size.toInt())
+
+/**
+ * Encodes this number using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
+ * while the highest bit indicates if more bytes are to come
+ */
+fun ULong.toAsn1VarInt(): ByteArray {
+    if (this < 128u) return byteArrayOf(this.toByte()) //Fast case
+    var offset = 0
+    var result= mutableListOf<Byte>()
+
+    var b0 = (this shr offset and 0x7FuL).toByte()
+    while ((this shr offset > 0uL) || offset == 0) {
+        result += b0
+        offset += 7
+        if (offset > (ULong.SIZE_BITS - 1)) break //End of Fahnenstange
+        b0 = (this shr offset and 0x7FuL).toByte()
+    }
+
+    return ByteArray(result.size){
+        result[result.size-1-it] or  (if (it < result.size-1) 0x80 else 0x00).toByte()
```

Comment:
How about refactoring to this:

```kotlin
    return with(result) {
        ByteArray(size) { fromBack(it) or asn1VarIntByteMask(it) }
    }
```

with helper functions

```kotlin
private fun MutableList<Byte>.asn1VarIntByteMask(it: Int) = (if (isLastIndex(it)) 0x00 else 0x80).toByte()

private fun MutableList<Byte>.isLastIndex(it: Int) = it == size - 1

private fun MutableList<Byte>.fromBack(it: Int) = this[size - 1 - it]

```

Reply:
isn't that a bit overkill?

Reply:
Just thought it would make at more readable

Reply:
done